Bottom: 4c0754349aa9a534958dfcb547d662ee21cccc21
Top:    453f5bac79bc10c4d0664076d2610bb1659a9bb1
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-12-09 23:08:11 +0000

hid/common: Fix nogui_prompt_for() and nogui_fileselect() in various ways:

1. These functions must return allocated memory, strdup will do nicely.

2. fgets will insert the newline character into the buffer, so we need
   to check if we just got a '\r' or '\n' as our first character, not
    just '\0' when deciding whether to return the default string or not.

3. DO NOT strcpy a the default string...
   we don't know if it will overflow our buffer

4. For the "fileselect" case, return NULL if the user didn't give us a
   filename, and the caller didn't specify a default string. Prompt for
   will return strdup (""), equivalent to what it previously did.


---

diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index 2b98a75..18bc637 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -294,14 +294,23 @@ static char *
 nogui_prompt_for (const char *msg, const char *default_string)
 {
   static char buf[1024];
+  char *s;
+
   if (default_string)
     printf ("%s [%s] : ", msg, default_string);
   else
     printf ("%s : ", msg);
-  fgets (buf, 1024, stdin);
-  if (buf[0] == 0 && default_string)
-    strcpy (buf, default_string);
-  return buf;
+
+  s = fgets (buf, 1024, stdin);
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_string != NULL)
+        return strdup (default_string);
+      else
+        return strdup ("");
+    }
+  else
+    return strdup (buf);
 }
 
 /* FIXME - this could use some enhancement to actually use the other
@@ -312,14 +321,23 @@ nogui_fileselect (const char *title, const char *descr,
 		  const char *history_tag, int flags)
 {
   static char buf[1024];
+  char *s;
+
   if (default_file)
     printf ("%s [%s] : ", title, default_file);
   else
     printf ("%s : ", title);
-  fgets (buf, 1024, stdin);
-  if (buf[0] == 0 && default_file)
-    strcpy (buf, default_file);
-  return buf;
+
+  s = fgets (buf, 1024, stdin);
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_file != NULL)
+        return strdup (default_file);
+      else
+        return NULL;
+    }
+  else
+    return strdup (buf);
 }
 
 static int
